Analytics: PageViewTracker and event properties

rodrigoi opened the pull request:
One of the limitation with the `PageViewTracker` component is that it does not support sending named properties along with the `path` and the `title`. That makes using `analytics.pageView.record` the only option when we need to extract parameters from the `url` or `state` to pass for paths like `/checkout/thank-you/:receiptId`.

This PR modifies the `recordPageView` action to add a properties argument that's added to the `payload`. It also changes the handling of that `payload` on the `middleware` from a list of allowed items to a list of blocked items. The assumption is that anything extra is a parameter of the `path`.

Looking for regressions, this change the way we send `client_id` along from https://github.com/Automattic/wp-calypso/pull/20502, so ideally the login page should be tested.

Let the discussion begin.

Pendings:

- [x] Add Tests

gwwar commented on client/state/analytics/middleware.js:
```diff
@@ -32,13 +32,18 @@ const eventServices = {
+// list of unsafe params that need to be blocked from beign passed down to the recorder.
+const PAGE_VIEW_SERVICES_BLOCKED_PARAMS = [ 'service' ];
+
+const omitUnsafeParams = flowRight(
```
Unit tests would be much appreciated here. Should we also warn/error for any camel case properties?

gwwar replied:
It appears that service is only used for the switch below, we can remove that value earlier. (Though we should audit this with each function in the switch.)
```jsx
const { service = 'default', params } = payload;

switch ( type ) {
 			case ANALYTICS_EVENT_RECORD:
 				return invoke( eventServices, service, payload );	 				return invoke( eventServices, service, payload );
//...
```

rodrigoi replied:
Updated on 658415952601d26c02d344a04ee684cc08ce6548.

Adding unit tests.

gwwar commented on client/state/analytics/middleware.js:
```diff
@@ -32,13 +32,18 @@ const eventServices = {
-// Whitelists specific parameters to avoid polluting page view events
-const PAGE_VIEW_SERVICES_ALLOWED_PARAMS = [ 'client_id' ];
+// list of unsafe params that need to be blocked from beign passed down to the recorder.
+const PAGE_VIEW_SERVICES_BLOCKED_PARAMS = [ 'service' ];
```
On that note, are there other reserved values we should not override for analytics reporting? Blacklists are typically more dangerous than whitelists since we need knowledge about everything we should block vs knowing what's safe.

gwwar replied:
I dug a bit further to see where client_id was added in https://github.com/Automattic/wp-calypso/pull/20502 

Overall I think we're okay with this or further destructuring before this method, since a plain tracks call will blindly report values 👍

gwwar commented on client/lib/analytics/page-view-tracker/index.jsx:
```diff
@@ -68,7 +76,7 @@ export class PageViewTracker extends React.Component {
 
 		if ( ! delay ) {
-			return recorder( path, title );
+			return recorder( path, title, undefined, properties );
```
`undefined` looks super odd here. 

Maybe a sign of a code smell. Since service is optional, we may fold this into the last properties argument. We can leave this for a future refactor though. For now, the two allowed values are `default` which records to ga+tracks, or just `ga`. So let's pass through `'default'` over undefined.

rodrigoi replied:
Updated on a102071268633b91e8ebd2345867b6d6ad7ecc3f.

Ideally I would like to change the order of the arguments to leave the service last, but haven't checked all the places where `record` is used. Will be janitorialized later.

gwwar commented:
Nice, I like the approach here! I did some light testing, and I didn't notice any odd behaviors.

Let's see what would recommendations look like? We can start with this being manual at first. Say if folks normalize `:domain` in a page url, the page view tracker component expects a prop value with name of `domain`?

Copons commented:
> Let's see what would recommendations look like? We can start with this being manual at first. Say if folks normalize `:domain` in a page url, the page view tracker component expects a prop value with name of `domain`?

I'd leave this without any actual recommendations because the requirements might vary wildly between sections or even between routes.

It looks to me that the majority of page views recordings don't actually need additional parameters, so for me this is more about getting 1:1 parity with the direct `analytics.record.pageView` call (and being able to work on `/checkout` 🙂) than adding a widely used feature.

rodrigoi commented:
Thanks for the reviews!

This is a very naive approach, extracted from an experiment to fix the `/checkout` routes, and this definitely need tests and there are things to be fixed. But the idea is to validate the use case for the component and get potential high level problems with this addition.

> so for me this is more about getting 1:1 parity with the direct `analytics.record.pageView`

that's absolutely right. The motivation behind this is to eventually have a single `PageViewTracker` on each layout and simplify the controllers to make easier to work with routes.
A good first step is to remove all the `analytics.*` calls from the controllers, and then we could potentially move to a single tracker per section. So feature parity with `analytics.record.pageView` is a requirement for that.

We may need to work with routing and state (to remove the `sectionifyWithRoutes` madness for example), but the idea is to make the first of very small steps towards a simpler architecture for routing/tracking.

gwwar commented:
So I suppose last items to be done before 🚢 would be either to pull out `service` earlier so we don't need `omitUnsafeParams` or add tests to `omitUnsafeParams`.

rodrigoi commented:
thanks @Automattic/lannister for the reviews!

I've introduced these changes on 8680507653f08db2c6da57533fde6d86229bf9e0:

- `payload` is now always a `{}`. Some actions were passing down a string.
- `middleware` tests now do a `calledWithExactly` check for all the parameters
- added a test for the only `reducer` action.
- added a missing test for `addHotJarScript`.

I'm confident I've checked all the use cases, but, 👀 are always welcome.

rodrigoi commented:
@Automattic/lannister updated the component docs. Proofreading is appreciated 📘 .

gwwar commented on client/state/analytics/middleware.js:
```diff
@@ -32,45 +32,45 @@ const eventServices = {
 
 			case ANALYTICS_TRACKS_ANONID_SET:
-				return analytics.tracks.setAnonymousUserId( payload );
+				return setAnonymousUserId( params );
```
Careful! The original payload param isn't an object.
```
export const setTracksAnonymousUserId = anonId => ( {
	type: ANALYTICS_TRACKS_ANONID_SET,
	meta: {
		analytics: [
			{
				type: ANALYTICS_TRACKS_ANONID_SET,
				payload: anonId,
			},
		],
	},
} );
```

gwwar replied:
It's not clean, but we'll want to pass through the original payload and keep the left side for these two.

For example:
```jsx
const { b = 'default' , ...rest } = 5678;
// b = default
// rest = {};
// we lose the original paylod
```

gwwar commented on client/state/analytics/middleware.js:
```diff
@@ -32,45 +32,45 @@ const eventServices = {
 
 			case ANALYTICS_TRACKS_OPT_OUT:
-				return analytics.tracks.setOptOut( payload );
+				return setOptOut( params );
```
```
export const loadTrackingTool = trackingTool => ( {
	type: ANALYTICS_TRACKING_ON,
	meta: {
		analytics: [
			{
				type: ANALYTICS_TRACKING_ON,
				payload: trackingTool,
			},
		],
	},
} );
```

gwwar commented on client/state/analytics/middleware.js:
```diff
@@ -32,45 +32,45 @@ const eventServices = {
 
+const setAnonymousUserId = ( { anonId } ) => analytics.tracks.setAnonymousUserId( anonId );
+
+const setOptOut = ( { isOptingOut } ) => analytics.tracks.setOptOut( isOptingOut );
```
We also won't need these helper methods unless we want to normalize the payload before going through the switch.

bperson replied:
I think those are used to transform the "object-payload" to a raw type since we cannot use raw types as `payload` anymore. I'm not entirely sure whether adding another layer is a good idea though: it might be worth updating the entire chain of calls to handle directly `{ isOptingOut }`. ( I'm not 100% sure but I think `setAnonymousUserId` is basically dead code at this point )

gwwar commented on client/state/analytics/actions.js:
```diff
@@ -85,7 +85,7 @@ export const loadTrackingTool = trackingTool => ( {
 			{
 				type: ANALYTICS_TRACKING_ON,
-				payload: trackingTool,
+				payload: { trackingTool },
```
Ah okay you also made updates here. Let's keep this small for now, and follow up We'd need to retest the privacy update settings otherwise

gwwar commented:
@bperson I'm kind of on the fence for updating the privacy actions here or leaving them alone for now. Do you mind helping to test cases, and let us know your thoughts.

gwwar commented:
Testing instructions should be available from https://github.com/Automattic/wp-calypso/pull/23231.

@rodrigoi feel free to dismiss my review if you want to back out the privacy refactors. Otherwise we'll want to confirm that things still work as expected.

gwwar commented on client/state/analytics/middleware.js:
```diff
@@ -32,45 +32,45 @@ const eventServices = {
-				return statBump( payload );
+				return statBump( params );
 
 			case ANALYTICS_TRACKING_ON:
```
We need steps for this one if we're going to change the payload signature.

gwwar replied:
Unfortunately there aren't testing instructions in https://github.com/Automattic/wp-calypso/pull/12332

rodrigoi commented:
Thanks @gwwar and @bperson for taking the time to review this!

Let me go item by item:

#### Changes on the Payload signature
I was very thorough when making those changes. The `payload` change is circumscribed only to the `state` functions, the actions are always called using their action creators and I followed a TDD approach, writing failing tests first. I've also retroactively added tests for the `reducer` and manually tested the new functions (for example _HotJar_ tracking on `/checklist/:site`). I'm pretty confident there should be no regressions, so I don't think we should remove this change.

#### Testing instructions on #21652
That PR got merged a month ago, and I think that testing instructions on a merged PR do not need to be 100% reliable because they are only accurate as long as the PR is open. I'll be more than happy to update that PR and any related open PRs when this lands as @bperson suggested on https://github.com/Automattic/wp-calypso/pull/23911#pullrequestreview-112345050

#### `setOptOut` and `setAnonymousUserId`
I've added those two functions to follow the same pattern established by `pageViewServices`, `loadTrackingTool` and `statBump`: calling a function that destructures the payload. I'm ok with destructuring on `setOptOut`, but I'd rather not change the signature of the methods in `analytics` for a design change in `state`.

#### `ANALYTICS_TRACKS_ANONID_SET`
This action is indeed never dispatched anywhere. I've created a janitorial PR to remove the dead code: #24208.

gwwar commented:
Sorry can't update the dismiss note. Used the wrong handle.

kwight commented on client/lib/analytics/page-view-tracker/README.md:
```diff
@@ -48,3 +55,26 @@ render() {
+import PageViewTracker from 'analytics/page-view-tracker';
+
+render() {
+	const { postId, postTile } = this.props;
```
`postTitle` 🙃

Copons replied:
⬛️⬜️⬛️⬜️⬛️⬜️⬛️⬜️
⬜️⬛️⬜️⬛️⬜️⬛️⬜️⬛️
⬛️⬜️⬛️⬜️⬛️⬜️⬛️⬜️
⬜️⬛️⬜️⬛️⬜️⬛️⬜️⬛️
⬛️⬜️⬛️⬜️⬛️⬜️⬛️⬜️
⬜️⬛️⬜️⬛️⬜️⬛️⬜️⬛️
⬛️⬜️⬛️⬜️⬛️⬜️⬛️⬜️

rodrigoi replied:
Fixed on c911a1298004be364521696566eed7984af180d0.

![giggle](https://media.tenor.com/images/0a209ffa8b996323d9b5aeb7712241e4/tenor.gif)

gwwar replied:
Nice catch ✨

kwight commented:
I'm unclear what and how I should be testing at this point 🙃 .

I've fired up the branch, roamed through a variety of different paths, and saw no issues. I've checked a purchase flow (adding a plan) to cover `/checkout/thank-you/:receiptId` et al., and that looked good.

I also went through the login process, but didn't see a `client_id` – is that correct? I didn't follow instructions from other PRs, I just opened http://calypso.localhost:3000/log-in with `debug` set and followed the steps: page views didn't get a `client_id` prop.

gwwar commented on client/state/analytics/middleware.js:
```diff
@@ -31,42 +31,42 @@ const eventServices = {
 
 const statBump = ( { group, name } ) => analytics.mc.bumpStat( group, name );
 
+const setAnonymousUserId = ( { anonId } ) => analytics.tracks.setAnonymousUserId( anonId );
```
This can be removed. Maybe double check your lint settings as the following should throw:
```
```
/home/circleci/wp-calypso/client/state/analytics/middleware.js
  47:7  error  'setAnonymousUserId' is assigned a value but never used  no-unused-vars
```

rodrigoi commented:
thanks @gwwar, @kwight, @Copons, @bperson for all the help getting this ready 🙂
